File: 0018_4Sum/4Sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}
def four_sum(nums, target)
  return [] if nums.length < 4
  nums.sort!
  result = []
  (0...nums.length - 3).each do |i|
    next if i > 0 && nums[i] == nums[i - 1]
    (i + 1...nums.length - 2).each do |j|
      next if j > i + 1 && nums[j] == nums[j - 1]
      k = j + 1
      l = nums.length - 1
      while k < l
        sum = nums[i] + nums[j] + nums[l] + nums[k]
        if target > sum
          k += 1
        elsif target < sum
          l -= 1
        else
          result << [nums[i], nums[j], nums[l], nums[k]]
          loop do
            k += 1
            break unless k < l && nums[k] == nums[k - 1]
          end
          loop do
            l -= 1
            break unless k < l && nums[l] == nums[l + 1]
          end
        end
      end
    end
  end
  result
end

File: 0018_4Sum/4Sum.py
# -*- coding: utf-8 -*-

"""
Given an array nums of n integers and an integer target, are there elements a, b, c, and d in nums such that a + b + c + d = target? Find all unique quadruplets in the array which gives the sum of target.

Note:

The solution set must not contain duplicate quadruplets.

Example:

Given array nums = [1, 0, -1, 0, -2, 2], and target = 0.

A solution set is:
[
  [-1,  0, 0, 1],
  [-2, -1, 1, 2],
  [-2,  0, 0, 2]
]
"""


class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        if len(nums) < 4:
            return []
        nums.sort()
        length, temp = len(nums), []
        for i in range(length - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            for j in range(i + 1, length - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                m = j + 1
                n = length - 1
                while m < n:
                    num = nums[i] + nums[j] + nums[n] + nums[m]
                    if num < target:
                        m += 1
                    elif num > target:
                        n -= 1
                    else:
                        temp.append([nums[i], nums[j], nums[n], nums[m]])
                        m += 1
                        n -= 1
        result = []
        [result.append(t) for t in temp if t not in result]
        return result
